test/continuity.lean

```diff
@@ -0,0 +1,42 @@
+import topology.tactic
+import topology.algebra.monoid
+import topology.instances.real
+import analysis.special_functions.trigonometric
+
+example {X Y : Type*} [topological_space X] [topological_space Y]
+  (f₁ f₂ : X → Y) (hf₁ : continuous f₁) (hf₂ : continuous f₂)
+  (g : Y → ℝ) (hg : continuous g) : continuous (λ x, (max (g (f₁ x)) (g (f₂ x))) + 1) :=
+by show_term { continuity }
+-- prints `refine ((continuous.comp hg hf₁).max (continuous.comp hg hf₂)).add continuous_const`
+
+example {κ ι : Type}
+  (K : κ → Type) [∀ k, topological_space (K k)] (I : ι → Type) [∀ i, topological_space (I i)]
+  (e : κ ≃ ι) (F : Π k, homeomorph (K k) (I (e k))) :
+  continuous (λ (f : Π k, K k) (i : ι), F (e.symm i) (f (e.symm i))) :=
+by show_term { continuity }
+-- prints, modulo a little bit of cleaning up the pretty printer:
+--   `exact continuous_pi (λ i, ((F (e.symm i)).continuous).comp (continuous_apply (e.symm i)))`
+
+open real
+
+local attribute [continuity] continuous_exp continuous_sin continuous_cos
+
+example : continuous (λ x : ℝ, exp ((max x (-x)) + sin x)^2) :=
+by show_term { continuity }
+-- prints
+--   `exact (continuous_exp.comp ((continuous_id.max continuous_id.neg).add continuous_sin)).pow 2`
+
+
+
+-- On the other hand, it's not that hard to get `continuity` to run off the rails:
+-- ```
+-- example : continuous (λ x : ℝ, exp ((max x (-x)) + sin (cos x))^2) :=
+-- by show_term { continuity }
+-- ```
+-- produces a giant term, and repeats of these goals:
+-- ⊢ continuous complex.re
+-- ⊢ continuous complex.exp
+-- ⊢ continuous coe
+-- ⊢ continuous complex.im
+
+-- Perhaps we should make `continuity` use `apply { md := reducible }` under the hood?
```

Comment:
Do we want to address this now or later?

Reply:
I think this is a good idea; too much unfolding is especially dangerous when it will make Lean start trying to unify functions. Unfortunately it looks like a fair bit of work to thread a reducibility option to where it needs to get.

Reply:
I did manage to do it, but one of my examples broke with a deterministic timeout, and I lost enthusiasm for fixing it.

I would be happy to rejigger `apply_rules` to take this option, but I think I'd prefer to do that as a separate PR to keep it simple. I'll try to do that in the next few days.